pkg/plugin/processor/builtin/unwrap_opencdc.go

```diff
@@ -0,0 +1,290 @@
+	ref, err := u.fieldRefRes.Resolve(&rec)
+	if err != nil {
+		return sdk.ErrorRecord{Error: cerrors.Errorf("failed resolving record reference: %w", err)}
+	}
+
+	var data opencdc.StructuredData
+	switch v := ref.Get().(type) {
+	case opencdc.RawData:
+		// unmarshal raw data to structured
+		if err := json.Unmarshal(v.Bytes(), &data); err != nil {
+			return sdk.ErrorRecord{Error: cerrors.Errorf("failed to unmarshal raw data as JSON: %w", err)}
+		}
+	case opencdc.StructuredData:
+		data = v
+	case nil:
+		return sdk.ErrorRecord{Error: cerrors.New("field to unmarshal is nil")}
+	default:
+		return sdk.ErrorRecord{Error: cerrors.Errorf("unexpected data type %T", v)}
+	}
+
+	opencdcRec, err := u.unmarshalRecord(data)
+	if err != nil {
+		return sdk.ErrorRecord{Error: cerrors.Errorf("failed unmarshaling record: %w", err)}
+	}
+	opencdcRec.Position = rec.Position
+
+	// Position is the only key we preserve from the original record to maintain the reference respect other messages
+	// that will be coming from in the event of chaining pipelines (e.g.: source -> kafka, kafka -> destination)
+	return sdk.SingleRecord(opencdcRec)
+}
+
+func (u *unwrapOpenCDC) Teardown(context.Context) error {
+	return nil
+}
+
+func (u *unwrapOpenCDC) unmarshalRecord(structData opencdc.StructuredData) (opencdc.Record, error) {
+	operation, err := u.unmarshalOperation(structData)
+	if err != nil {
+		return opencdc.Record{}, err
+	}
+
+	metadata, err := u.unmarshalMetadata(structData)
+	if err != nil {
+		return opencdc.Record{}, err
+	}
+
+	key, err := u.unmarshalKey(structData)
+	if err != nil {
+		return opencdc.Record{}, err
+	}
+
+	payload, err := u.unmarshalPayload(structData)
+	if err != nil {
+		return opencdc.Record{}, err
+	}
+
+	return opencdc.Record{
+		Key:       key,
+		Metadata:  metadata,
+		Payload:   payload,
+		Operation: operation,
+	}, nil
+}
+
+// unmarshalOperation extracts operation from a structuredData record.
+func (u *unwrapOpenCDC) unmarshalOperation(structData opencdc.StructuredData) (opencdc.Operation, error) {
+	var operation opencdc.Operation
+	op, ok := structData["operation"]
+	if !ok {
+		return operation, cerrors.Errorf("record payload after doesn't contain operation")
+	}
+
+	switch opType := op.(type) {
+	case opencdc.Operation:
+		operation = opType
+	case string:
+		if err := operation.UnmarshalText([]byte(opType)); err != nil {
+			return operation, cerrors.Errorf("invalid operation %q", opType)
+		}
+	default:
+		return operation, cerrors.Errorf("expected a opencdc.Operation or a string, got %T", opType)
+	}
+	return operation, nil
+}
+
+// unmarshalMetadata extracts metadata from a structuredData record.
+func (u *unwrapOpenCDC) unmarshalMetadata(structData opencdc.StructuredData) (opencdc.Metadata, error) {
+	var metadata opencdc.Metadata
+	meta, ok := structData["metadata"]
+	if !ok {
+		return metadata, cerrors.Errorf("record payload after doesn't contain metadata")
```

Comment:
I guess this error msg is a bit confusing, and it only mentions payload.after, but it can be something else

Reply:
I've changed this to just "no metadata", since the calling method already says "failed unmarshalling metadata". Let me know how it looks like now!